apps/examples/memory_fragmentation_test/memory_fragmentation_test.c

```diff
@@ -0,0 +1,318 @@
+	int i;
+
+	for (i = 0; i < MAX_SIZE_EXPONENT; ++i) {
+		num_free[i] = num[i];
+	}
+
+	for (i = 0; i < MAX_SIZE_EXPONENT; ++i) {
+		while (num_free[i]) {
+			temp = list[i].next;
+			if (temp) {
+				temp->prev->next = temp->next;
+				if (temp->next) {
+					temp->next->prev = temp->prev;
+				} 
+				free(temp->data);
+				free(temp);
+				--num_alloc[i];
+			}
+			--num_free[i];
+		}
+	}
+}
+
+static void print_usage(void)
+{
+	printf("Usage: memfrag f1 f2 r\n");
+	printf("       f1 and f2 should range [1-9].\n");
+	printf("       f1/10 and f2/10 of initial number of small and large memory allocation, respectively\n");
+	printf("       are released and allocated repeatedly r times.\n");
+	printf("       Initial number of each size memory allocation is given inside the source code.\n");
+	printf("       If you want to change this value, modify the value of the array called numof_size.\n");
+	exit(-1);
+}
+
+static void memory_fragmentation_test(int argc, char *argv[])
+{
+	int i;
+	int j;
+	int f1;
+	int f2;
+	int r;
+	int num_alloc[MAX_SIZE_EXPONENT] = {0, };
+	int num_alloc_tmp[MAX_SIZE_EXPONENT] = {0, };
+	int num_free[MAX_SIZE_EXPONENT] = {0, };
+	struct alloc_list list[MAX_SIZE_EXPONENT];
+	struct alloc_list *next[MAX_SIZE_EXPONENT];
+	struct alloc_list *item;
+
+	/* The number of memory segments whose size is 2^(i+4) where i is the index
+	 * of the arrary of numof_size[].
+	 * The following data was derived from the memory usage of the DA easy-setup
+	 * when MAX_SIZE_EXPONENT is 12.
+	 */
+	// DA easy-setup memory usage 
+	//int numof_size[MAX_SIZE_EXPONENT] = {57, 809, 561, 74, 57, 90, 17, 8, 11, 15, 4, 9};
+	// Three small memory sizes and two large memory sizes
+	int numof_size[MAX_SIZE_EXPONENT] = {700, 300, 200, 0, 0, 0, 0, 0, 70, 50, 0, 0}; 
+
+	for (i = 0; i < MAX_SIZE_EXPONENT; ++i) {
+		list[i].data = NULL;
+		list[i].next = NULL;
+		list[i].prev = NULL;
+	}
+
+	if (argc < 4) {
+		print_usage();
+	}
+
+	f1 = strtol(argv[2], (char **)NULL, 10);
+	if ((f1 < 1) || (f1 > 9)) {
+		print_usage();
+	}
+
+	f2 = strtol(argv[3], (char **)NULL, 10);
+	if ((f2 < 1) || (f2 > 9)) {
+		print_usage();
+	}
+
+	r = strtol(argv[4], (char **)NULL, 10);
+	if (r < 0) {
+		print_usage();
+	}
+
+	srand(SEED);
+
+	/* Allocate memory according to 'numof_size' */
+	memory_allocation(list, numof_size, num_alloc);
+
+	/* Free the number of memory allocation according to 'num_free' 
+	 * with a first-come first-free manner.
+	 */
+	printf("Test Configuration\n");
+	printf("Size [bytes]	# of init alloc		# of free\n");
+	for (i = 0; i < NUMOF_SMALL_MEMORY; ++i) {
+		num_free[i] = num_alloc[i] * f1 / 10;
+		if (num_alloc[i] > 0) {
+			printf("%d		%d			%d\n", 1 << (i + 4), num_alloc[i], num_free[i]);
+		}
+	}
+	for (i = NUMOF_SMALL_MEMORY; i < MAX_SIZE_EXPONENT; ++i) {
+		num_free[i] = num_alloc[i] * f2 / 10;
+		if (num_alloc[i] > 0) {
+			printf("%d		%d			%d\n", 1 << (i + 4), num_alloc[i], num_free[i]);
+		}
+	}
+	memory_free(num_free, num_alloc, list);
```

Comment:
We should verify this as well and TC should fail if memory_free fails

Reply:
I added the checkpoint logic.